src/include/server/mir/observer_multiplexer.h

```diff
@@ -199,17 +211,24 @@ void ObserverMultiplexer<Observer>::for_each_observer(MemberFn f, Args&&... args
         std::is_member_function_pointer<MemberFn>::value,
         "f must be of type (Observer::*)(Args...), a pointer to an Observer member function.");
     auto const invokable_mem_fn = std::mem_fn(f);
-    std::shared_lock<decltype(observer_mutex)> lock{observer_mutex};
-    for (auto& observer_pair: observers)
+    decltype(observers) local_observers;
     {
-        observer_pair.first.spawn(
-            [invokable_mem_fn, weak_observer = observer_pair.second, args...]() mutable
-            {
-                if (auto observer = weak_observer->lock())
+        std::lock_guard<decltype(observer_mutex)> lock{observer_mutex};
+        local_observers = observers;
+    }
+    for (auto& weak_observer: local_observers)
+    {
+        if (auto observer = weak_observer->lock())
```

Comment:
I've not written a test (yet) but I'm *pretty* sure that this no longer upholds the “once `unregister_interest()` returns there is guaranteed to be no thread in or entering this Observer”.

If thread *A* is currently in `for_each_observer()` and the dispatch calls `unregister_interest(foo)` while thread *B* is also in `for_each_observer()` in the `for(... : local_observers)` loop, then thread *B* can hit `if (auto observer = weak_observer->lock())` with `*weak_observer == foo`.

At this point, thread *B* will block waiting to acquire the lock until thread *A* finishes the `unregister_interest(foo)` call.

Now *B* will *successfully* acquire a lock of `weak_observer`, because the `lock()` call has kept a strong reference to the observer around, and `observer.spawn()` will (if the executor is an immediate executor) now call into the Observer `foo`.

I *think* this could be fixed by switching the locking around in `WeakObserver::lock()`, but this is an annoyingly subtle area of code.

Reply:
Yes, there were a few functions where `WeakObserver`'s observer wasn't being protected by the mutex. In the updated PR I lock the mutex earlier in `WeakObserver::lock()` (as you suggested), and I also do the reset process in `WeakObserver::maybe_reset()` (so it can keep the mutex locked for the duration).